web_src/js/features/gitgraph.js

```diff
@@ -1,13 +1,579 @@
+
+
+// Although inspired by the https://github.com/bluef/gitgraph.js/blob/master/gitgraph.js
+// this has been completely rewritten with almost no remaining code
+
+// GitGraphCanvas is a canvas for drawing gitgraphs on to
+class GitGraphCanvas {
+  constructor(canvas, widthUnits, heightUnits, config) {
+    this.ctx = canvas.getContext('2d');
+
+    const width = widthUnits * config.unitSize;
+    this.height = heightUnits * config.unitSize;
+
+    const devicePixelRatio = window.devicePixelRatio || 1;
+    const backingStoreRatio =
+      this.ctx.webkitBackingStorePixelRatio ||
+      this.ctx.mozBackingStorePixelRatio ||
+      this.ctx.msBackingStorePixelRatio ||
+      this.ctx.oBackingStorePixelRatio ||
+      this.ctx.backingStorePixelRatio ||
+      1;
+
+    const ratio = devicePixelRatio / backingStoreRatio;
```

Comment:
```suggestion
    const ratio = window.devicePixelRatio || 1;
```

`backingStoreRatio` is deprecated and always `1` on all browsers so we can simplify this. See https://stackoverflow.com/questions/24332639